### DISCUSSION TITLE
What are the expectations for a recycle implementation?

### DISCUSSION THREAD

**Discussion by Electron100** [bikeshedder/deadpool#371] [Q&A] [answered]
I'm unclear from the documentation and example what the expectations are for an implementation of `deadpool::managed::Manager::recycle`. The example in the docs simply no-ops, and the docstring just says "Tries to recycle an instance of `[Manager::Type]`" without defining what it means to recycle.

Is a `recycle` implementation simply expected to check that the connection is still valid and error if it's not? Anything else?

Thanks in advance.

**Comment by bikeshedder** [comment, marked as answer]
I somehow assumed that it's pretty self explanatory what it is expected to do and didn't put too much thought into writing more documentation for it.

- `create` is used to create new objects
- `recycle` is used to recycle existing objects

The important bit missing here is...

> `recycle` **MUST** ensure that the object is in working condition. If the object was modified (e.g. listeners registered, connection parameters changed, etc.) it **SHOULD** be reset or discarded by returning an error.

The reset step is not a hard requirement. It might actually be the case that you know for sure that your code doesn't modify the objects. Performing a full cleanup can hurt performance. e.g. that's the reason why `deadpool-postgres` does come with a configurable `RecyclingMethod` and makes the user of the library choose what recycling actually does.

https://docs.rs/deadpool-postgres/latest/deadpool_postgres/enum.RecyclingMethod.html
